Document wasm JS builtins (#37201)
mdn/content: files/en-us/webassembly/javascript_builtins/index.md

---
title: WebAssembly JavaScript builtins
slug: WebAssembly/JavaScript_builtins
page-type: guide
---

{{WebAssemblySidebar}}

WebAssembly JavaScript builtins are wasm equivalents of JavaScript operations that provide a way to use JavaScript features inside wasm modules without having to import JavaScript glue code to provide a bridge between JavaScript and WebAssembly values and calling conventions.

This provides performance improvements — importing glue code for primitives such as {{jsxref("String")}} can come with a significant overhead. WebAssembly and most languages that target it expect a tight sequence of inline operations rather than an indirect function call, which is how regular imported functions work.

This article explains how builtins work and which ones are available, then provides a usage example.

## Problems with importing JavaScript functions

> There are several problems with importing functions from JavaScript to WebAssembly modules:

Author: OK, rewritten as 

> There are several reasons why importing functions from JavaScript to WebAssembly modules introduce significant performance overheads:

- Existing APIs require a conversion to handle differences around the [`this`](/en-US/docs/Web/JavaScript/Reference/Operators/this) value, which WebAssembly function `import` calls leave as `undefined`.
- Certain primitives use JavaScript operators such as [`===`](/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality) and [`<`](/en-US/docs/Web/JavaScript/Reference/Operators/Less_than) that cannot be imported.
- Most JavaScript functions are extremely permissive of the types of values they accept, and it's desirable to leverage WebAssembly's type system to remove those checks and coercions wherever we can.

> Considering these problems, creating built-in definitions that adapt existing JavaScript primitives to WebAssembly is simpler, more flexible, and better for performance than importing them.

Reviewer: Not sure where you see increased flexibility, but the built-in definitions are certainly good for performance.

Author: OK, I've removed the flexibility part. It now reads"

> Considering these problems, creating built-in definitions that adapt existing JavaScript primitives to WebAssembly is simpler and better for performance than importing them.

> ## WebAssembly JavaScript builtin types

Reviewer: "## WebAssembly JavaScript Builtins"

Author: I've changed this to "## Available WebAssembly JavaScript builtins": I wanted to be clear that here we are discussing which builtins are available (which is what led me to use "types" in the first place ;-) ), rather than just discussing builtins in general.

> The first builtin types to be implemented are {{jsxref("String")}} operations. This is because the most pressing use case is languages that would benefit from using the JavaScript `String` type to implement their strings. You can find a list of the `String` operations with builtin equivalents defined for them at [JS String Builtin API](https://github.com/WebAssembly/js-string-builtins/blob/main/proposals/js-string-builtins/Overview.md#js-string-builtin-api).

Reviewer: We're not implementing builtin _types_, but _operations_. As far as WebAssembly types are concerned, every imported string is just a generic `externref`.

Author: OK, understood. I've gone through the pages and made sure to find other ways to refer to "different available builtins" than "builtin types", given that "type" has a specific meaning that may confuse.

> Other primitive types are likely to be supported via builtins in the future.

Reviewer: "Other builtins are likely to be supported in the future." 

It's not about types, and not about the JS concept of "primitive" types/values. For example, one of the next candidates could be `DataView`-related operations, but that's totally TBD (i.e. probably too early to mention here).

Author: Good call; I've used your wording here. I've also scanned through the pages and made sure that there are no other problematic uses of "primitives" remaining.

Builtins are enabled at compile-time by specifying the `builtins` property during compilation. This goes inside the `compileOptions` object, and its value is an array of the builtin types you want to enable:

```js
WebAssembly.compile(bytes, { builtins: ["js-string"] });
```

The `compileOptions` object is available to the following functions:

- [`WebAssembly.compile()`](/en-US/docs/WebAssembly/JavaScript_interface/compile_static)
- [`WebAssembly.compileStreaming()`](/en-US/docs/WebAssembly/JavaScript_interface/compileStreaming_static)
- [`WebAssembly.instantiate()`](/en-US/docs/WebAssembly/JavaScript_interface/instantiate_static)
- [`WebAssembly.instantiateStreaming()`](/en-US/docs/WebAssembly/JavaScript_interface/instantiateStreaming_static)
- [`WebAssembly.validate()`](/en-US/docs/WebAssembly/JavaScript_interface/validate_static)
- The [`WebAssembly.Module()`](/en-US/docs/WebAssembly/JavaScript_interface/Module/Module) constructor

> You can also include the `importedStringConstants` property inside `compileOptions`, which enables and specifies an identifier for imported global string constants to be used by the builtins:

Reviewer: Below (line 104), you use the term "namespace". Might be good to use the same term here, to avoid confusion. Perhaps phrase it roughly like: "..., which selects one import namespace for imported global string constants that the engine will populate automatically".

(You can use these imported string constants for anything you want, not just for the imported JS String Builtins. For example, you could import additional custom functions that work on strings.)

Author: As before, I've fixed this to use "namespace". I've also updated the text clarify that the imported global string constants are a separate feature and not part of builtins, but I've also said that they are often used together, to justify them being in this article.

Ideally, we'd probably have a separate article for them, but I am unsure how to structure this. It would probably end up really small. So, we'll leave it as-is for now.

```js
WebAssembly.compile(bytes, {
  builtins: ["js-string"],
  importedStringConstants: "#",
});
```

> [!NOTE]

> > The above example uses `"#"` as the imported global identifier for illustrative purposes. In production, however, it is best practice to use the empty string to save on module file size. The identifier is repeated for every string literal, and real-world modules can have tens of thousands of them, so the saving can be significant.

Reviewer: If you switch to the term "namespace" above, then please also switch here.

Might also be worth mentioning that the choice is made by the authors of the toolchain that will generate the Wasm modules. Once you have a `.wasm` file and want to embed it in your JavaScript, you can't freely choose this namespace any more; you have to use what that `.wasm` file expects.

Author: OK, I've updated "identifier" to namespace in the different pages, and taken heed of your suggested wording.

Author: I've now also added the note about the namespace choice being made by the toolchain authors. I've added it to the bottom of the note.

This article explains how builtins work and which ones are available, then provides a usage example.

## Problems with importing JavaScript functions

There are several reasons why importing functions from JavaScript to WebAssembly modules introduces significant performance overheads:

- Existing APIs require a conversion to handle differences around the [`this`](/en-US/docs/Web/JavaScript/Reference/Operators/this) value, which WebAssembly function `import` calls leave as `undefined`.
- Certain primitives use JavaScript operators such as [`===`](/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality) and [`<`](/en-US/docs/Web/JavaScript/Reference/Operators/Less_than) that cannot be imported.
- Most JavaScript functions are extremely permissive of the types of values they accept, and it's desirable to leverage WebAssembly's type system to remove those checks and coercions wherever possible.

> Considering these problems, creating built-in definitions that adapt existing JavaScript functionality to WebAssembly is simpler and better for performance than importing it.

Reviewer 2: One thing I took away from the explainer, which is not captured IMO here, is that this is not for the whole web platform, or even for the whole of JavaScript. Generally you will import web APIs you need as this is efficient enough. This is for a few primitives where that need to be optimized/inlined, or similar.

Specifically see this bit "However, the overhead of importing glue code is prohibitive for primitives such as [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String), [ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer), [RegExp](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp), [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map), and [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) where the desired overhead of operations is a tight sequence of inline instructions, not an indirect function call (which is typical of imported functions)."

Make sense? My concern is that people will wonder why arbitrary API of choice is not made importable.

Author: OK, that makes sense. I've taken what was the second paragraph out of the intro (making the intro short and sweet) and moved it down to the "Problems with importing JavaScript functions" section. I've then rearranged the section a bit, adding in the explanation I think you want, to make it clear why builtins are useful for specific functionality only.

Let me know what you think.

## Available WebAssembly JavaScript builtins

> The first builtin operations to be implemented are {{jsxref("String")}} operations. This is because the most pressing use case is languages that would benefit from using the JavaScript `String` type to implement their strings. You can find a list of the `String` operations with builtin equivalents defined for them at [JS String Builtin API](https://github.com/WebAssembly/js-string-builtins/blob/main/proposals/js-string-builtins/Overview.md#js-string-builtin-api).

Reviewer 2: No intent to add MDN documents for those? It would be good to at least list them to provide a flavour for the options here. If they were documented, where would they "live" - e.g. under `Globals`?

Reviewer: They're not globals. I think they'd have to be in a new section.

Author: I wasn't intending to document the Wasm code for the actual builtins; they aren't part of the web platform, as such. In the same way, we don't provide detailed guides or reference docs for the Wasm text format. I guess what I could do is just list the JS functions with builtin equivalents. I'll update the section to do that, and we'll see what we think about it.

Reviewer 2: The changes address my point, which is making it very obvious what features are supported - so thanks.

> , we don't provide detailed guides or reference docs for the Wasm text format

FWIW Partially done in https://developer.mozilla.org/en-US/docs/WebAssembly/Reference :-)
We don't publish the limited info available but I doubt there is any conscious decision that this is out of scope for MDN.

Author: Oh wow, I did not know that existed! It looks unfinished and doesn't appear to be linked from anywhere. Another job for another day, methinks.

> The first builtin operations to be implemented are {{jsxref("String")}} operations. This is because the most pressing use case is languages that would benefit from using the JavaScript `String` type to implement their strings. You can find a list of the `String` operations with builtin equivalents defined for them at [JS String Builtin API](https://github.com/WebAssembly/js-string-builtins/blob/main/proposals/js-string-builtins/Overview.md#js-string-builtin-api).
> 
> Other builtins are likely to be supported in the future.

Reviewer 2: I would also be tempted to do this something like below, in order to ease future extensibility. Also because I don't know how useful it is to say the reason why strings were picked.

```suggestion
The following builtin APIs are implemented:

- `js-string`: A library of {{jsxref("String")}} operations. You can find a list of the `String` operations with builtin equivalents defined for them at [JS String Builtin API](https://github.com/WebAssembly/js-string-builtins/blob/main/proposals/js-string-builtins/Overview.md#js-string-builtin-api).

Note that {{jsxref("String")}} operation have been implemented first but more are likely to be supported in the future.
```

Reviewer 2: PS If you do list them as I suggest above then consider having a `js-string` heading so that as more libraries are added we can easily see where to add specific docs about what is available.

Author: I've done something similar but used subheadings rather than bullets. Let me know what you think.

Reviewer 2: Yes, I like what you did more.

## How do you use builtins?

> Builtins work in a similar way to functions imported from JavaScript, except that you are using standard wasm function equivalents for performing JavaScript operations that are defined in a reserved namespace (`wasm:`). This being the case, browsers can predict and generate optimal code for them. This section summarizes how to use them.

Reviewer 3: ```suggestion
Builtins work in a similar way to functions imported from JavaScript, except that you are using standard Wasm function equivalents for performing JavaScript operations that are defined in a reserved namespace (`wasm:`). This being the case, browsers can predict and generate optimal code for them. This section summarizes how to use them.
```

### JavaScript API features

> Builtins are enabled at compile-time by specifying the `builtins` property during compilation. This goes inside the `compileOptions` object, and its value is an array of the builtins you want to enable:

Reviewer 2: ```suggestion
Builtins are enabled at compile-time by specifying the `compileOptions.builtins` property as an argument when calling method to compile or intantiate a module. Its value is an array of the builtins you want to enable:
```

I'd like to say "builtin libraries" - what name is given to each set? Point is to make it clear it is a bunch of things, not just one

Reviewer: I don't think there is a term yet, as this concept is only just being established. "builtin libraries" sounds good to me. "sets of builtins" could be an alternative.

Author: I thought about this a bit, and ended up going with:

> Builtins are enabled at compile-time by specifying the `compileOptions.builtins` property as an argument when calling methods to compile and/or instantiate a module. Its value is an array of strings that identify the sets of builtins you want to enable:

"libraries" is ok, but I feel like it has too many connotations around "JavaScript libraries", which might give some people the wrong idea.

Reviewer 2: So they are "sets" :-). The text looks good.

```js
WebAssembly.compile(bytes, { builtins: ["js-string"] });
```

The `compileOptions` object is available to the following functions:

- [`WebAssembly.compile()`](/en-US/docs/WebAssembly/JavaScript_interface/compile_static)
- [`WebAssembly.compileStreaming()`](/en-US/docs/WebAssembly/JavaScript_interface/compileStreaming_static)
- [`WebAssembly.instantiate()`](/en-US/docs/WebAssembly/JavaScript_interface/instantiate_static)
- [`WebAssembly.instantiateStreaming()`](/en-US/docs/WebAssembly/JavaScript_interface/instantiateStreaming_static)
- [`WebAssembly.validate()`](/en-US/docs/WebAssembly/JavaScript_interface/validate_static)
- The [`WebAssembly.Module()`](/en-US/docs/WebAssembly/JavaScript_interface/Module/Module) constructor

If you want to use imported global string constants with your builtins, you can also include the `importedStringConstants` property inside `compileOptions`. This selects an import namespace for imported global string constants that the engine will populate automatically:

```js
WebAssembly.compile(bytes, {
  builtins: ["js-string"],
  importedStringConstants: "#",
});
```

> You don't have to use imported global string constants along with builtins; they are two separate features. Howeever, they are often used together.

Reviewer 3: ```suggestion
You don't have to use imported global string constants along with builtins; they are two separate features. However, they are often used together.
```

> [!NOTE]

> > The above example uses `"#"` as the imported global string namespace for illustrative purposes. In production, however, it is best practice to use the empty string to save on module file size. The namespace is repeated for every string literal, and real-world modules can have tens of thousands of them, so the saving can be significant. This choice is made by the authors of the toolchain that will generate the wasm modules. Once you have a `.wasm` file and want to embed it in your JavaScript, you can't freely choose this namespace any more; you have to use what the `.wasm` file expects.

Reviewer 3: ```suggestion
> The above example uses `"#"` as the imported global string namespace for illustrative purposes. In production, however, it is best practice to use the empty string to save on module file size. The namespace is repeated for every string literal, and real-world modules can have tens of thousands of them, so the saving can be significant. This choice is made by the authors of the toolchain that will generate the Wasm modules. Once you have a `.wasm` file and want to embed it in your JavaScript, you can't freely choose this namespace any more; you have to use what the `.wasm` file expects.
```

> > The above example uses `"#"` as the imported global string namespace for illustrative purposes. In production, however, it is best practice to use the empty string to save on module file size. The namespace is repeated for every string literal, and real-world modules can have tens of thousands of them, so the saving can be significant. This choice is made by the authors of the toolchain that will generate the Wasm modules. Once you have a `.wasm` file and want to embed it in your JavaScript, you can't freely choose this namespace any more; you have to use what the `.wasm` file expects.

Reviewer 2: As part of https://github.com/mdn/content/pull/37201/files#r1887811889 we need to explain why this is needed (at all). Why a namespace?

Reviewer: Because _every_ Wasm import has a namespace.

Author: I think this is covered by the extra explanation in the separate imported global string constants article. Let me know if you feel it needs any more.

Reviewer 2: > Because every Wasm import has a namespace.

I've left a comment in https://github.com/mdn/content/pull/37201#discussion_r1893482947 explaining why this was confusing to me. Essentially, you call it an import, but the strings are defined _right there_ in the module - from the point of view of the ignorant (me) that's not an import just because you put a cheeky `import` statement at the front :-)

Silliness aside, a few words might be useful, and possibly a link to where it says all imports have a namespace and why that is so.

I leave it up to you to decide if needed  because it is probable that I am too far from your straw-man audience to be a good judge on this.

Author: See other comment. I think the first section in the new article helps to make it clear why this is an import, for those who are not sure.

### WebAssembly module features

> Over in your webassembly module, you can now import string literals, specifying the same namespace you specified in `importedStringConstants` over in the JavaScript:

Reviewer 3: ```suggestion
Over in your WebAssembly module, you can now import string literals, specifying the same namespace you specified in `importedStringConstants` over in the JavaScript:
```

```wasm
(global $h (import "#" "hello ") externref)
(global $w (import "#" "world!") externref)
```

You can also import builtins as specified in the `compileOptions` object from the `wasm:` namespace (in this case, the [`concat()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/concat) function; see also the [equivalent built-in definition](https://github.com/WebAssembly/js-string-builtins/blob/main/proposals/js-string-builtins/Overview.md#wasmjs-string-concat)):

```wasm
(func $concat (import "wasm:js-string" "concat")
    (param externref externref) (result (ref extern)))

> ```

Reviewer 2: I would split all the stuff about the global string constants into its own section - after a all, as you say it is an independent part.

The main reason for this though is that it make no sense, and you need some space to describe it.
- What is a global string constant?
- It looks like you import them like this `import "#" "hello "` - where are they imported from? I mean this looks like a **declaration** of a string not some kind of import. 
- Why is the `#` needed? I see that this is the namespace declared in JS side, but what "reality" does it have in the JS side?

Reviewer: In case it's helpful:

The Wasm module phrases string constants as regular imported globals. They _could_ indeed just be imported as regular globals (which is exactly what polyfills do). The `importedStringConstants` compile-time argument is what triggers the magic: it instructs the engine to automatically satisfy these imports (using their names as their value), so that they don't need to be provided explicitly.

The `#` is needed because every Wasm import has a "namespace" and a "name". This two-level naming also comes in really handy, because it makes it possible to select one namespace as the group of "magic" imports that will automatically be satisfied with strings by the engine. The choice of `#` as this namespace is arbitrary, as long as the Wasm module and its embedding JavaScript code are in sync. The example here could alternatively use `"string constants"` or whatever; in production we recommend the empty string to save module size (as mentioned elsewhere in this PR).

Perhaps this mechanism is best explained by retracing the path of ideas that led to this design: in the state before JS Builtins were standardized, a Wasm module could import JS strings like any other JS object, for example:
```
(global (import "string_constants" "string_constant_1") externref)
(global (import "string_constants" "string_constant_2") externref)
```
The embedding JS would then provide them like so:
```
importsObject = {
  ...,
  string_constants: {
    string_constant_1: "hello",
    string_constant_2: "world",
    ...
  },
};
... = WebAssembly.instantiate(..., importsObject);
```
But this is not great for several reasons:
- Download size is large (much larger than the strings themselves), because for every string, you need the definition of the imported global in the Wasm module, and the definition of the value on the JS side.
- All these bytes also take time to parse before the Wasm module can be instantiated.
- For Wasm module optimizers, it's very inconvenient to have to modify an accompanying JS file along with the Wasm module, e.g. when optimizing out unused string constants.
- For debugging, it's nice when the name of an import at least hints at the value that will be imported there. Since import names can be arbitrary unicode strings, it's possible to take the _entire_ string as the import's name. That yields:
```
(global (import "string_constants" "hello") externref)
(global (import "string_constants" "world") externref)
```
in the Wasm module and
```
  string_constants: {
    hello: "hello",
    world: "world",
    ...
  },
```
on the JS side. 
And then it's easy to see that the JS-side list `hello: "hello", world: "world", foo: "foo", bar: "bar", guesswhat: "guesswhat", ...` is so "boring" that it's easy to automate it away. So that's where the new `importedStringConstants: "string_constants"` flag comes in: it tells the Wasm engine "just look at all the imported globals in the `string_constants` namespace in the Wasm module, and for every `foo` in there, create a string `"foo"` for it, don't make me enumerate them all in JS".

Author: OK, ok, this took a while ;-)

But I have updated the PR (in my next commit) to provide a separate imported string constants guide, which incorporates the relevant stuff that was in the builtins guide, and @jakobkummerow's wonderful explainer content above (thanks for that, so helpful!)

You should probably both have a look at this when its available. I'll poke you for re-reviews once my changes are pushed.

Reviewer 2: Thank you for this. I read @chrisdavidmills doc first and then saw this when going through the email chain. It is all helpful.

> ## Builtins example

Reviewer 2: IMO Example is really clear.

Author: Nice! ;-)

> Let's work through a basic but complete example to show how builtins are used. This example will define a function inside a wasm module that concatenates two strings together and prints the result to the console, then export it. We will then call the exported function from JavaScript.

Reviewer 3: ```suggestion
Let's work through a basic but complete example to show how builtins are used. This example will define a function inside a Wasm module that concatenates two strings together and prints the result to the console, then export it. We will then call the exported function from JavaScript.
```

The example we'll be referring to uses the [`WebAssembly.instantiate()`](/en-US/docs/WebAssembly/JavaScript_interface/instantiate_static) function on the webpage to handle the compilation and instantiation; you can find this and other examples on our `webassembly-examples` repo — see [`js-builtin-examples`](https://github.com/mdn/webassembly-examples/tree/main/js-builtin-examples).

> You can build up the example by following the steps below; also [see it running live](https://mdn.github.io/webassembly-examples/js-builtin-examples/instantiate/).

Reviewer 2: I was kind of hoping for "Hello World" when I clicked the link:

![image](https://github.com/user-attachments/assets/10d6c301-cbff-4cec-bebd-55e6138d4550)

Author: It is, but the hello world is in the console ;-)

To make this clearer, I've updated this sentence to

> You can build up the example by following the steps below. In addition, you can [see it running live](https://mdn.github.io/webassembly-examples/js-builtin-examples/instantiate/) — open your browser's JavaScript console to see the example output.

I've also opened up a separate PR on the `webassembly-examples` repo to improve the HTML content on those examples. "My template" is not very helpful!

> ### The JavaScript

Reviewer 2: ```suggestion
### JavaScript
```

The JavaScript for the example is shown below. To test this locally, include it in an HTML page using a method of your chooosing (for example, inside {{htmlelement("script")}} tags, or in an external `.js` file referenced via `<script src="">`).

```js
const importObject = {
  // Regular import
  m: {
    log: console.log,
  },
};

const compileOptions = {
  builtins: ["js-string"], // Enable JavaScript string builtins
  importedStringConstants: "#", // Enable imported global string constants
};

fetch("log-concat.wasm")
  .then((response) => response.arrayBuffer())
  .then((bytes) => WebAssembly.instantiate(bytes, importObject, compileOptions))
  .then((result) => result.instance.exports.main());
```

The JavaScript:

> - Defines an `importObject` that specifies a function `"log"` at a namespace `"m"` to import into the wasm module during instantiation. It's the {{domxref("console.log()")}} function.

Reviewer 3: ```suggestion
- Defines an `importObject` that specifies a function `"log"` at a namespace `"m"` to import into the Wasm module during instantiation. It's the {{domxref("console.log()")}} function.
```

> - Defines a `compileOptions` function that includes the `builtins` and `importedStringConstants` properties. As discussed earlier, this is needed to enable builts and imported global string constants.

Reviewer 3: ```suggestion
- Defines a `compileOptions` function that includes the `builtins` and `importedStringConstants` properties. As discussed earlier, this is needed to enable builtins and imported global string constants.
```

> - Defines a `compileOptions` function that includes the `builtins` and `importedStringConstants` properties. As discussed earlier, this is needed to enable builtins and imported global string constants.

Reviewer 2: ```suggestion
- Defines a `compileOptions` object that includes the `builtins` and `importedStringConstants` properties. As discussed earlier, this is needed to enable builtins and imported global string constants.
```

- Uses {{domxref("Window.fetch", "fetch()")}} to fetch the wasm module (`log-concat.wasm`), converts the response to an {{jsxref("ArrayBuffer")}} using {{domxref("Response.arrayBuffer")}}, then compiles and instantiates the wasm module using [`WebAssembly.instantiate()`](/en-US/docs/WebAssembly/JavaScript_interface/instantiate_static).

> - Calls the `main()` function exported from the wasm module.

Reviewer 3: ```suggestion
- Calls the `main()` function exported from the Wasm module.
```

> ### The wasm module

Reviewer 3: ```suggestion
### The Wasm module
```

> ### The Wasm module

Reviewer 2: ```suggestion
### Wasm module
```

Specifically, importing functions from JavaScript to WebAssembly modules creates performance problems for the following reasons:

- Existing APIs require a conversion to handle differences around the [`this`](/en-US/docs/Web/JavaScript/Reference/Operators/this) value, which WebAssembly function `import` calls leave as `undefined`.
- Certain primitives use JavaScript operators such as [`===`](/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality) and [`<`](/en-US/docs/Web/JavaScript/Reference/Operators/Less_than) that cannot be imported.
- Most JavaScript functions are extremely permissive of the types of values they accept, and it's desirable to leverage WebAssembly's type system to remove those checks and coercions wherever possible.

Considering these problems, creating built-in definitions that adapt existing JavaScript functionality such as {{jsxref("String")}} primitives to WebAssembly is simpler and better for performance than importing it and relying on indirect function calls.

## Available WebAssembly JavaScript builtins

The below sections detail the available builtins. Other builtins are likely to be supported in the future.

### String operations

> The {{jsxref("String")}} operations with builtin equivalents are:

Author: Cool, thanks for the guidance here. I thought we should include these other ones, but I wasn't quite sure how to approach it.

I've added these, and also added a definition for `"wasm:js-string" "equals"`, as it was missing too.

- {{jsxref("String.fromCharCode()")}} (`"wasm:js-string" "fromCharCode"`)
- {{jsxref("String.fromCodePoint()")}} (`"wasm:js-string" "fromCodePoint"`)
- {{jsxref("String.prototype.charCodeAt()")}} (`"wasm:js-string" "charCodeAt"`)
- {{jsxref("String.prototype.codePointAt()")}} (`"wasm:js-string" "codePointAt"`)
- {{jsxref("String.prototype.length")}} (`"wasm:js-string" "length"`)
- {{jsxref("String.prototype.concat()")}} (`"wasm:js-string" "concat"`)
- {{jsxref("String.prototype.substring()")}} (`"wasm:js-string" "substring"`)

> - {{jsxref("String.prototype.localeCompare()")}} (`"wasm:js-string" "compare"`)

Reviewer: Not quite accurate, depending on (default) locale. For example, in many locales including at least "en" and "de", `"ö".localeCompare("z") === -1`, but `"ö" < "z" === false`.

Author: Ah, ok, that makes sense. I've rewritten the description of this one to describe it in terms of determining order of two strings — less than, greater than, or equal.

## Available WebAssembly JavaScript builtins

The below sections detail the available builtins. Other builtins are likely to be supported in the future.

### String operations

The available {{jsxref("String")}} builtins are:

- `"wasm:js-string" "cast"`

  - : Throws an error if the provided value is not a string. Roughly equivalent to:

    ```js
    if (typeof obj !== "string") throw new WebAssembly.RuntimeError();
    ```

- `"wasm:js-string" "compare"`
  - : Compares two string values and determines their order. Returns `-1` if the first string is [less than](/en-US/docs/Web/JavaScript/Reference/Operators/Less_than) the second, `1` if the first string is [greater than](/en-US/docs/Web/JavaScript/Reference/Operators/Greater_than) the second, and `0` if the strings are [strictly equal](/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality).
- `"wasm:js-string" "concat"`
  - : Equivalent to {{jsxref("String.prototype.concat()")}}.
- `"wasm:js-string" "charCodeAt"`
  - : Equivalent to {{jsxref("String.prototype.charCodeAt()")}}.
- `"wasm:js-string" "codePointAt"`
  - : Equivalent to {{jsxref("String.prototype.codePointAt()")}}.
- `"wasm:js-string" "equals"`

>   - : Compares two string values for [strict equality](/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality), returning `1` if they are equal, and `0` if not.

Reviewer: Might be worth pointing out here that `"equals"` is the only function of the bunch that doesn't throw/trap for `null` inputs, so Wasm modules don't need to separately check for non-`null`-ness before calling `"equals"`. (All the other functions have no reasonable way to handle `null` inputs, so they throw for them.)

You can then try validating this module with the [validate()](/en-US/docs/WebAssembly/JavaScript_interface/validate_static) method, but note how the result is negated with the `!` operator — remember that builtins are supported if the module is _invalid_:

```js
const compileOptions = {
  builtins: ["js-string"],
};

fetch("module.wasm")
  .then((response) => response.arrayBuffer())
  .then((bytes) => WebAssembly.validate(bytes, compileOptions))
  .then((result) => console.log(`Builtins available: ${!result}`));
```

The above module code is so short that you could just validate the literal bytes rather than downloading the module. A feature detection function could look like so:

```js
function JsStringBuiltinsSupported() {
  let bytes = new Uint8Array([
    0, 97, 115, 109, 1, 0, 0, 0, 1, 4, 1, 96, 0, 0, 2, 23, 1, 14, 119, 97, 115,
    109, 58, 106, 115, 45, 115, 116, 114, 105, 110, 103, 4, 99, 97, 115, 116, 0,
    0,
  ]);
  return !WebAssembly.validate(bytes, { builtins: ["js-string"] });
}
```

> [!NOTE]

> > In many cases it is not necessary to feature detect builtins. Another option could be to provide regular imports alongside the builtins, and supporting browsers will just ignore the fallbacks.

Reviewer: I suppose this is where we should add a link to the polyfill (in a separate PR next year).

Author: Yes indeed!

Reviewer 2: By "In many cases it is not necessary to feature detect builtins" I think you are actually saying "there are alternatives to feature checking" - right? If so, this is a little ambiguous - it could mean you don't have to because it doesn't matter if they work or not.

So maybe 

> There are alternatives to feature checking for builtin support. For example, you can polyfil or .xxxx ...

- Certain primitives use JavaScript operators such as [`===`](/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality) and [`<`](/en-US/docs/Web/JavaScript/Reference/Operators/Less_than) that cannot be imported.
- Most JavaScript functions are extremely permissive of the types of values they accept, and it's desirable to leverage WebAssembly's type system to remove those checks and coercions wherever possible.

Considering these problems, creating built-in definitions that adapt existing JavaScript functionality such as {{jsxref("String")}} primitives to WebAssembly is simpler and better for performance than importing it and relying on indirect function calls.

## Available WebAssembly JavaScript builtins

The below sections detail the available builtins. Other builtins are likely to be supported in the future.

### String operations

The available {{jsxref("String")}} builtins are:

- `"wasm:js-string" "cast"`

  - : Throws an error if the provided value is not a string. Roughly equivalent to:

    ```js
    if (typeof obj !== "string") throw new WebAssembly.RuntimeError();
    ```

> - `"wasm:js-string" "compare"`

Reviewer 2: Are the definitions such as https://github.com/WebAssembly/js-string-builtins/blob/main/proposals/js-string-builtins/Overview.md#wasmjs-string-cast useful? i.e. do they add anything over what you have put below for readers?|

If so, then I would link them individually like this:

```suggestion
- [`"wasm:js-string" "compare"`](https://github.com/WebAssembly/js-string-builtins/blob/main/proposals/js-string-builtins/Overview.md#wasmjs-string-cast)
```

Author: I've tried it out, and I like it. I think it is quite useful, plus it allows me to get rid of the line below the DL.
